### PR TITLE
bugfix: request duration reflect the time to establish connection

### PR DESCRIPTION
**Pull request by jizhuozhi**
### Issues associated with this PR

#2266 

### Solutions
Record request start time before getting connection.

### CODE REVIEW

**Comment by doujiang24** [comment]
Cool! I do think it's a problem. But, I'm not sure if it's a proper fix.
A bit worried about the backward compatibility.

**Comment by nejisama** [comment]
The NewStream of ConnPool is established asynchronously, and a NewStream may establish up to 3 connections at a time.
When the next request is made to try NewStream, it may use the asynchronously established connection of the previous request. At this time, the time consumption is relatively low, but it does not mean that the corresponding target machine network is better than the previous one.

**Comment by jizhuozhi** [comment]
Hello, @doujiang24 and @nejisama .

My intention is not to accurately represent the quality of service of the network, but to reflect the complete request processing time (that is, the time perceived by the end user).

In scenarios where pools are used or protocols that support multiplexing, only initializing the connection will bring additional overhead. After the connection is initialized, it should take the same time as the business request processing time. But when the quality of service of the network drops, the connection will never be added to the pool due to the connection timeout, so the connection will be continuously established. At this time, the delay will increase up to the connection timeout time, but it cannot be found in the metrics.

Or should a new metric be created to reflect this duration?

**Comment by nejisama** [comment]
Maybe we should add a new metrics, and count it at real connect

**Comment by jizhuozhi** [comment]
> Maybe we should add a new metrics, and count it at real connect

How about `UpstreamProxyDuration`? Because it's really proxied to upstream ~~

File: pkg/metrics/upstream.go [hunk 1]
```diff
@@ -45,11 +45,13 @@ const (
 	UpstreamRequestPendingOverflow                 = "request_pending_overflow"
 	UpstreamRequestDuration                        = "request_duration_time"
 	UpstreamRequestDurationTotal                   = "request_duration_time_total"
+	UpstreamProxyDuration                          = "proxy_duration_time"
```
Comment (doujiang24) [inline pkg/metrics/upstream.go:48]: better add some comments.

**Review by doujiang24 — COMMENTED** [review]
a small comment, otherwise, LGTM.

**Comment by jizhuozhi** [comment]
The meaning of `UpstreamRequestDuration` has changed. In a word:
- `UpstreamRequestDuration`: The time from the initiation to the end of the request, including internally obtaining available connections and other processing

- `UpstreamProxyDuration`: Same meaning as the original `UpstreamRequestDuration`, only includes the processing time from writing the request to receiving the response

For Ops, the request-response interval should completely represent the complete time from the initiation to the end of the request, including obtaining the available connection and sending the request data until receiving the response from the server. But now `UpstreamRequestDuration` does not include the available connection time, which will bring costs for Ops: why there is such a big gap between the response time seen from the view of mosn and the view of end users.

To fill this gap, we revised the meaning of `UpstreamRequestDuration` and introduced a new `UpstreamProxyDuration` to replace the original `UpstreamRequestDuration`.

**Comment by jizhuozhi** [comment]
cc @rayowang

File: pkg/metrics/upstream.go (hunk 1 above)
Comment (spacewander) [inline pkg/metrics/upstream.go:48]: Could you add some comments to show the difference between request/proxy duration?

**Pull request closed without merging**